source/slang-glslang/slang-glslang.cpp

```diff
@@ -190,6 +192,44 @@ static int glslang_dissassembleSPIRV(glslang_CompileRequest* request)
     return 0;
 }
 
+// We need a per process initialization
+class ProcessInitializer
+{
+public:
+    ProcessInitializer()
+    {
+        m_isInitialized = false;
+    }
+
+    bool init()
+    {
+        std::lock_guard<std::mutex> guard(m_mutex);
+        if (!m_isInitialized)
+        {
+            if (!glslang::InitializeProcess())
+            {
+                return false;
+            }
+            m_isInitialized = true;
+        }
+        return true;
+    }
+
+    ~ProcessInitializer()
+    {
+        // We *assume* will only be called once dll is detatched and that will be on a single thread
+        if (m_isInitialized)
+        {
+            glslang::FinalizeProcess();
+        }
+    }
+
+    std::mutex m_mutex;
+    bool m_isInitialized = false;
+};
+
+static ProcessInitializer s_processInitializer;
```

Comment:
I will strongly advocate that this is a *global* variable and thus should be `gProcessInitializer`, and that we should reserve the `s_` prefix (if we use it at all...) for variables that use `static` scoping to indicate they are effectively global but scoped within a `struct` or `class` type.

The underlying problem is that C++ uses `static` to mean too many different things at once.

Reply:
A more substantive question is: why does this use a mutex in the `init` method instead of just relying on a constructor?

I assume the answer is some combination of:
* Running in a separate `init` method guarantees we only pay for any overhead in glslang when it is actually used. This would matter if/when we statically link in `slang-glslang` instead of loading it dynamically.
* There are platform-specific considerations around shared library loading and constructor execution for C++ `static` variables, and avoiding reliance on the exact semantics helps us.

The one interesting bit here is that we haven't relied heavily on C++ stdlib types elsewhere in the code, but for stuff like threading and mutexes it seems like the simplicity of just using the stdlib is welcome compared to rolling out own wrapper types for this.

Reply:
Main reason for 'init' function is the 'how to return a result from ctor' problem. It's not globally initialized, to avoid the cost of setup when not used.

Reply:
I wouldn't be keen on 'gProcessInitializer', because it looks a little ugly, and is seemingly inconsistant with other conventions we do have. But if the point is 'it's an s and that means static' you could use `g_processInitializer`. I read the s_ as global - not 'it's a static I wonder what it means'. Admittedly it's origins are probably around static.